File: src/main/java/io/github/gravetii/dictionary/Trie.java
package io.github.gravetii.dictionary;

import io.github.gravetii.util.Alphabet;

public class Trie {
  private final TrieNode root;

  Trie() {
    this.root = new TrieNode();
  }

  /**
   * Add a new word to this trie.
   *
   * @param word the word to add
   */
  public void insert(String word) {
    TrieNode node = root;
    int score = 0;
    for (char ch : word.toCharArray()) {
      int idx = ch - 'a';
      Alphabet alphabet = Alphabet.values()[idx];
      TrieNode[] children = node.getChildren();
      if (children[idx] == null) children[idx] = new TrieNode();
      node = children[idx];
      score += alphabet.getScore();
    }

    node.setScore(score);
  }

  /**
   * Search for the given word in the trie and return its score.
   *
   * @param word the input word
   * @return the score of the word: > 0 if the word exists, = 0 if the word exists as a prefix, < 0
   *     if the word doesn't exist
   */
  public int search(String word) {
    TrieNode node = root;
    for (char ch : word.toCharArray()) {
      int idx = ch - 'a';
      TrieNode[] children = node.getChildren();
      if (children[idx] == null) return -1;
      node = children[idx];
    }

    return node != null ? node.getScore() : -1;
  }
}

File: src/main/kotlin/io/github/gravetii/dictionary/Trie.kt
package io.github.gravetii.dictionary

import io.github.gravetii.model.Alphabet

class TrieNode {
    val children: MutableMap<Int, TrieNode> = mutableMapOf()
    var score: Int = 0
}

class Trie {

    private val root = TrieNode()

    fun insert(word: String) {
        fun insertInternal(itr: CharIterator, root: TrieNode, score: Int) {
            val idx = itr.nextChar() - 'a'
            val alphabet = Alphabet.values()[idx]
            val node = root.children.getOrPut(idx) { TrieNode() }
            if (itr.hasNext()) insertInternal(itr, node, score + alphabet.score)
            else node.score = score + alphabet.score
        }

        insertInternal(word.iterator(), root, 0)
    }

    fun search(word: String): Int? {
        fun searchInternal(itr: CharIterator, root: TrieNode): Int? {
            val idx = itr.nextChar() - 'a'
            val node = root.children[idx]
            return if (itr.hasNext()) node?.let { searchInternal(itr, it) }
            else node?.score
        }

        return searchInternal(word.iterator(), root)
    }

}